File: mobile/lib/features/labtests/domain/lab_tests.dart
/// A test report the patient uploaded against a doctor-advised test.
class LabResult {
  const LabResult({
    required this.id,
    required this.testName,
    required this.note,
    this.photoUrl,
    this.createdAt,
    this.mimeType,
    this.originalName,
    this.sizeBytes,
    this.analysisStatus,
    this.analysisSummary,
    this.abnormal = const [],
  });

  final String id;
  final String testName;
  final String note;
  final String? photoUrl;
  final DateTime? createdAt;

  /// The stored file's type. `photoUrl` is a historical name — the upload sheet
  /// has always accepted PDFs and Office files too, and without this the screen
  /// drew every one of them as an image and got a broken thumbnail.
  final String? mimeType;
  final String? originalName;
  final int? sizeBytes;

  /// `pending` | `done` | `failed` | `unsupported`. Sent alongside the values
  /// so a report that could not be read says so, instead of looking like a
  /// report with nothing on it.
  final String? analysisStatus;
  final String? analysisSummary;

  /// Values the report itself marked out of range.
  final List<String> abnormal;

  bool get isReading => analysisStatus == 'pending';
  bool get couldNotRead =>
      analysisStatus == 'failed' || analysisStatus == 'unsupported';

  /// True when the report is something that can actually be shown as a picture.
  /// Unknown types are treated as documents: a file card that opens is a better
  /// wrong guess than an image box that cannot load.
  bool get isImage => mimeType?.startsWith('image/') ?? false;

  bool get hasFile => photoUrl != null && photoUrl!.isNotEmpty;

  factory LabResult.fromJson(Map<String, dynamic> j) => LabResult(
    id: j['id']?.toString() ?? '',
    testName: j['testName']?.toString() ?? '',
    note: j['note']?.toString() ?? '',
    photoUrl:
        (j['photoUrl'] == null || j['photoUrl'].toString().isEmpty)
            ? null
            : j['photoUrl'].toString(),
    createdAt: DateTime.tryParse(j['createdAt']?.toString() ?? '')?.toLocal(),
    mimeType: j['mimeType']?.toString(),
    originalName: j['originalName']?.toString(),
    sizeBytes: (j['sizeBytes'] as num?)?.toInt(),
    analysisStatus:
        (j['analysis'] as Map<String, dynamic>?)?['status']?.toString(),
    analysisSummary:
        (j['analysis'] as Map<String, dynamic>?)?['summary']?.toString(),
    abnormal:
        ((j['analysis'] as Map<String, dynamic>?)?['abnormal'] as List?)
            ?.map((e) => e.toString())
            .toList() ??
        const [],
  );
}

/// The tests the doctor advised + the reports the patient has uploaded.
class LabTestsView {
  const LabTestsView({
    required this.advised,
    required this.results,
    this.reportedNames,
  });

  final List<String> advised;
  final List<LabResult> results;

  /// The advised tests the server says already have a report, verbatim as they
  /// appear in [advised].
  ///
  /// The decision belongs on the server: it is the same question the doctor's
  /// panel and the upload reminder answer, and three different answers to it
  /// is what had a patient nudged to upload a report they had already sent.
  /// Null when the server did not send the field at all — an older build.
  /// An *empty* set is a real answer ("none of them are reported yet") and
  /// must not be mistaken for silence.
  final Set<String>? reportedNames;

  /// True once a report has been uploaded for [test].
  bool hasResultFor(String test) {
    final fromServer = reportedNames;
    if (fromServer != null) return fromServer.contains(test);
    // An older server that does not send the per-test status yet. Exact match
    // is weak — it misses "Vitamin D" against a lab's "Vitamin D (25-Hydroxy)"
    // — but a wrong tick is worse than a missing one, so it stays strict.
    return results.any((r) => r.testName.toLowerCase() == test.toLowerCase());
  }

  factory LabTestsView.fromJson(Map<String, dynamic> j) => LabTestsView(
    advised:
        (j['advised'] as List?)?.map((e) => e.toString()).toList() ?? const [],
    reportedNames:
        (j['advisedStatus'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .where((e) => e['reported'] == true)
            .map((e) => e['name'].toString())
            .toSet(),
    results:
        (j['results'] as List?)
            ?.whereType<Map<String, dynamic>>()
            .map(LabResult.fromJson)
            .toList() ??
        const [],
  );
}

File: backend/src/routes/labtests.js
import { Router } from 'express';
import mongoose from 'mongoose';
import { z } from 'zod';
import { requireAuth, resolvePatientScope } from '../middleware/auth.js';
import { validate } from '../middleware/validate.js';
import { asyncHandler, notFound } from '../middleware/errors.js';
import { audit } from '../middleware/audit.js';
import { LabResult } from '../models/LabResult.js';
import { Prescription } from '../models/Prescription.js';
import { analyseLabResult } from '../services/ai/labReport.js';
import { buildAnalytes } from '../services/analyteCatalog.js';
import { withdrawLabResult } from '../services/readingCorrections.js';
import { RECORD_STATE } from '../models/plugins/clinicalRecord.js';
import { recomputePatientRisk } from '../services/analytics.js';
import { reportedNames, isReported } from '../utils/testNames.js';
import { recordWindow, requireRecordAccess } from '../middleware/authorise.js';
import { attachableAssetId } from '../services/mediaAccess.js';

/**
 * The patient's lab tests: the tests the doctor advised (pulled from active
 * prescriptions) and the reports the patient has uploaded against them.
 * Mounted at /patients/:patientId/lab-tests.
 */
const router = Router({ mergeParams: true });
// Whose patient this is, then what this person may do with them: lab results and HbA1c.
// `resolvePatientScope` answers the first and was, until now, the only thing
// asked — so EDIT_RECORD was granted by every preset and enforced by nothing.
router.use(requireAuth, resolvePatientScope, requireRecordAccess());

function serialiseResult(r) {
  // `photo` is populated, so a report that is a PDF can say so. The field name
  // is historical — the upload sheet has always offered documents too, and the
  // client drew every one of them as an image, which is why a PDF report showed
  // a broken-picture icon.
  const asset = r.photo && typeof r.photo === 'object' ? r.photo : null;
  const photoId = asset ? asset._id : r.photo;
  return {
    id: String(r._id),
    testName: r.testName,
    note: r.note ?? '',
    photoUrl: photoId ? `/api/v1/uploads/${photoId}/raw` : null,
    mimeType: asset?.mimeType ?? null,
    originalName: asset?.originalName ?? null,
    sizeBytes: asset?.sizeBytes ?? null,
    // What was read off the page. `status` is sent as well as the values so a
    // report that could not be parsed reads as "needs a human", not as a
    // report with nothing on it.
    analysis: r.analysis?.status
      ? {
          status: r.analysis.status,
          summary: r.analysis.summary ?? null,
          hba1cPercent: r.analysis.hba1cPercent ?? null,
          fastingGlucoseMgDl: r.analysis.fastingGlucoseMgDl ?? null,
          postPrandialGlucoseMgDl: r.analysis.postPrandialGlucoseMgDl ?? null,
          ldl: r.analysis.ldl ?? null,
          hdl: r.analysis.hdl ?? null,
          triglycerides: r.analysis.triglycerides ?? null,
          creatinine: r.analysis.creatinine ?? null,
          testedOn: r.analysis.testedOn ?? null,
          abnormal: r.analysis.abnormal ?? [],
          // Uniform value/range/flag list built from the extracted numbers.
          analytes: buildAnalytes(r.analysis),
        }
      : null,
    createdAt: r.createdAt,
  };
}

router.get(
  '/',
  audit('read', 'LabResult'),
  asyncHandler(async (req, res) => {
    const [prescriptions, results] = await Promise.all([
      // Bounded like every other prescription read. What a practice advised
      // before this one was enrolled is that practice's clinical reasoning, and
      // "Vitamin D, HbA1c, lipid profile" is a diagnosis said out loud.
      Prescription.find({
        patient: req.patientId,
        isActive: true,
        ...recordWindow(req, 'issuedOn'),
      })
        .select('labTestsAdvised')
        .lean(),
      LabResult.find({
        patient: req.patientId,
        // `createdAt`, not `analysis.testedOn`. The window asks whether this
        // record existed during the practice's relationship with the patient,
        // and upload time answers that; the printed test date can be six weeks
        // earlier, and is missing entirely on a report the parser could not read
        // — a filter on it would hide those from everybody.
        ...recordWindow(req, 'createdAt'),
      })
        .sort({ createdAt: -1 })
        .limit(100)
        .populate('photo', 'mimeType originalName sizeBytes')
        .lean(),
    ]);
    const advised = [...new Set(prescriptions.flatMap((p) => p.labTestsAdvised ?? []).filter(Boolean))];

    // Whether each advised test has a report, decided here rather than in the
    // app. The client used to compare the two names with `==`, so "Vitamin D"
    // and the lab's "Vitamin D (25-Hydroxy)" were different tests and an
    // uploaded report never ticked off the advice it answered.
    //
    // `advised` keeps its old shape alongside this: a patient running an older
    // build still gets the list it expects.
    const reported = reportedNames(results);
    const advisedStatus = advised.map((name) => ({
      name,
      reported: isReported(name, reported),
    }));

    res.json({ advised, advisedStatus, results: results.map(serialiseResult) });
  }),
);

router.post(
  '/',
  validate({
    body: z
      .object({
        testName: z.string().trim().min(1).max(200),
        note: z.string().trim().max(1000).optional().default(''),
        photo: z.string().optional(),
      })
      .refine((b) => Boolean(b.photo) || b.note.trim().length > 0, { message: 'Add a photo or a note', path: ['photo'] }),
  }),
  audit('create', 'LabResult'),
  asyncHandler(async (req, res) => {
    // The patient's own report. The reader transcribes it into this patient's
    // HbA1c and glucose history, so somebody else's file here would move this
    // patient's numbers.
    const photo = await attachableAssetId(req.body.photo || undefined, {
      patientId: req.patientId,
      uploaderIds: [req.user._id],
    });
    const entry = await LabResult.create({
      patient: req.patientId,
      testName: req.body.testName,
      note: req.body.note,
      photo: photo || undefined,
    });
    // Populated before serialising so the row the client renders right after
    // upload knows whether it is a picture or a PDF, exactly as a reloaded one
    // does. Without it the first render of a freshly uploaded PDF is a broken
    // thumbnail that only fixes itself on the next refresh.
    await entry.populate('photo', 'mimeType originalName sizeBytes');

    // Read in the background. The patient sees their report listed at once
    // rather than waiting on a vision round-trip, and every failure path
    // leaves the report itself in place with a status saying what happened.
    if (entry.photo) {
      analyseLabResult(entry._id).catch(() => {});
    }

    res.status(201).json({ result: serialiseResult(entry.toObject()) });
  }),
);

/**
 * Withdraws a report uploaded by mistake.
 *
 * The values read off it go too. A wrong report that has already been
 * transcribed has moved this patient's HbA1c history, their glucose trend and
 * therefore their risk band — withdrawing only the report would leave the
 * numbers behind, still driving what all three panels show.
 *
 * Withdrawn, not deleted, and only what this report created: the old delete
 * removed every clinic glucose reading sharing its value and date, including
 * one the desk typed from the same printout. See services/readingCorrections.js.
 */
router.delete(
  '/:id',
  validate({ body: z.object({ reason: z.string().trim().min(3).max(300).optional() }).optional() }),
  audit('update', 'LabResult'),
  asyncHandler(async (req, res) => {
    if (!mongoose.isValidObjectId(req.params.id)) throw notFound('Report not found');
    const inReach = { _id: req.params.id, patient: req.patientId, ...recordWindow(req, 'createdAt') };
    const entry = await LabResult.findOne(inReach);
    if (!entry) {
      // Already withdrawn: a retry of the same removal gets the same answer.
      const withdrawn = await LabResult.exists({ ...inReach, recordState: RECORD_STATE.VOIDED });
      if (withdrawn) return res.status(204).end();
      throw notFound('Report not found');
    }

    await withdrawLabResult({
      entry,
      by: req.user._id,
      byRole: req.user.role,
      reason: req.body?.reason ?? (req.user.role === 'patient' ? 'Removed by the patient.' : 'Removed by the clinic.'),
    });

    // The record this fed has changed, so the band computed from it must be
    // recomputed — otherwise a withdrawn report leaves the patient sitting in a
    // risk band nothing on their record supports any more.
    recomputePatientRisk(req.patientId).catch(() => {});

    res.status(204).end();
  }),
);

export default router;
